feat(lsp): `root_markers` can control priority

TheRealLorenz opened the pull request:
Hello everyone! This PR is mostly intended to gather opinion on the topic.

While deprecating [`lspconfig`](https://github.com/neovim/nvim-lspconfig)`.util` module I came across an incoherent behaviour:

- `lspconfig.util.root_pattern` prioritises folder-up search
- `root_markers` of `vim.lsp.config` matches every pattern in every folder before going up (see `vim.fs.root()`.

With this modification now `root_markers` accepts `string[]` and `string[][]` (it actually also supports `string` and `function(name: string, path: string)`, but they're not documented), where each entry of the outer table is feed into `vim.fs.root()`. So:

```lua
{
  ...
  root_markers = { ".stylua.toml", ".git" }
  ...
}
```

Will look for `.stylua.toml` in every folder going up, if it isn't found it would do the same for `.git`. (Same behaviour as `util.root_pattern`, would make the change not affecting ported configs)

Instead:

```lua
{
  ...
  root_markers = { { ".stylua.toml", ".luarc.json" }, { ".git "} }
  ...
}
```

Would create two groups of "different priorites", where the it would look for `.stylua.toml` and `.luarc.json` (in this order) in every folder going up, if they aren't found it would do the same for `.git`.


This modification is not only "backward-compatible" with [nvim-lspconfig](https://github.com/neovim/nvim-lspconfig), but it offers the possibility to have different precedences in matching, as the "right way" is currently debated, it depends on how some languages are implemented and how some users work.

Reference:

> > I'm not entirely sure if this migration is ok. It seems like `util.root_pattern` and `vim.fs.root` behave differently
>
> See also https://github.com/neovim/nvim-lspconfig/pull/3651 . Deciding on well-defined behavior will require some thought.
>
> _Originally posted by @justinmk in https://github.com/neovim/nvim-lspconfig/issues/3711#issuecomment-2802304532_

See also:
https://github.com/neovim/nvim-lspconfig/pull/3651 
https://github.com/neovim/nvim-lspconfig/pull/2885

TheRealLorenz committed:
feat(lsp): `root_markers` folder-up first search

lewis6991 commented:
> Will look for .stylua.toml in every folder going up, if it isn't found it would do the same for .git. (Same behaviour as util.root_pattern, would make the change not affecting ported configs)

I think this is undesirable for when `root_markers` is `string[]`, and instead every string should have the same priority. Then to make ,`.stylua.toml` have higher priority you specify:

```lua
root_markers = {{' stylua.toml'},{'.git'}}
```

cbackas commented:
Does having these groups with different "priorities" solve the problem I was trying to solve with https://github.com/neovim/nvim-lspconfig/pull/3730 ? ([discussion which describes the problem](https://github.com/neovim/neovim/discussions/33473))

TheRealLorenz commented on runtime/lua/vim/lsp.lua:
```diff
@@ -651,7 +651,18 @@ function lsp.start(config, opts)
+        end
+      end
+    else
+      config.root_dir = vim.fs.root(bufnr, opts._root_markers)
```
It would be possible, but the signature of the function would become quite ugly in my opinion and it would probably be more breaking?

lewis6991 replied:
I think out of scope for this PR.

TheRealLorenz commented:
> Does having these groups with different "priorities" solve the problem I was trying to solve with [neovim/nvim-lspconfig#3730](https://github.com/neovim/nvim-lspconfig/pull/3730) ? ([discussion which describes the problem](https://github.com/neovim/neovim/discussions/33473))

This would probably solve it (but it's not meant to address only that), but removing `.git` is still a good idea I think.

cbackas commented:
> > Does having these groups with different "priorities" solve the problem I was trying to solve with [neovim/nvim-lspconfig#3730](https://github.com/neovim/nvim-lspconfig/pull/3730) ? ([discussion which describes the problem](https://github.com/neovim/neovim/discussions/33473))
> 
> This would probably solve it (but it's not meant to address only that), but removing `.git` is still a good idea I think.

👍🏻 Ok cool yeah I still think removing it is a good idea too since its easier to merge it in than it is to get it out (maybe impossible?)

lewis6991 commented:
Don't we also need to change the `vim.lsp.Config` type?

MariaSolOs commented:
I think it would also be helpful to document somewhere in `lsp.txt` the new behavior (maybe with the example from this PR's description).

TheRealLorenz committed:
docs(lsp): add documentation for `root_markers` priority

TheRealLorenz commented:
Edites vim.lsp.Config type and updated lsp.txt. I've never edited a vimdoc, hope I haven't screw up much.

Just for the sake of clarity I changed the wording from "markers precedence" to "markers priority"

lewis6991 commented on runtime/doc/lsp.txt:
```diff
@@ -228,6 +229,27 @@ To see the capabilities for a given server, try this in a LSP-enabled buffer: >v
+
+Example (mixed priorities):
+
+    root_markers = { { "stylua.toml", ".luarc.json" }, { ".git" } }
```
```suggestion
    root_markers = { { 'stylua.toml', '.luarc.json' }, '.git' }
```

This show that each element can be a `string[]` or a `string`.

TheRealLorenz replied:
Should I commit a fixup?

lewis6991 replied:
Can do, but will need to squash before we merge.

lewis6991 commented:
Also this **needs** a news entry.

TheRealLorenz commented on runtime/doc/lsp.txt:
```diff
@@ -228,6 +229,27 @@ To see the capabilities for a given server, try this in a LSP-enabled buffer: >v
+================================================================================
+ROOT MARKERS                                            *lsp-root-markers*
+
+In `vim.lsp.Config.root_markers` it's possibile to specify marker's
```
I honestly didn't know if it could fit into the `@field` section.

lewis6991 replied:
Should be fine, everything above `@field` will be used and can be markdown.

TheRealLorenz replied:
Done

TheRealLorenz committed:
docs(news): add new `root-markers` functionality

Co-authored-by: Maria José Solano <majosolano99@gmail.com>

Co-authored-by: Gregory Anders <github@gpanders.com>

Co-authored-by: Justin M. Keyes <justinkz@gmail.com>

lewis6991 commented on runtime/doc/news.txt:
```diff
@@ -70,7 +70,8 @@ HIGHLIGHTS
 
-• todo
+• `root_markers` in |vim.lsp.Config| can now specify markers priority
+  (defaults to folder-up first search).
```
Need to explain the breaking change.

lewis6991 replied:
```suggestion
  (defaults to directory-up first search).
```

lewis6991 commented on runtime/doc/news.txt:
```diff
@@ -70,7 +70,10 @@ HIGHLIGHTS
+• `root_markers` in |vim.lsp.Config| can now specify markers priority.
+  Defaults to directory-up first search, which is more inline with
+  `nvim-lspconfig` but changes the default behaviour, which looked for every
+  marker before going direcotory-up.
```
Can you provide examples to make this extra clear? This file gets a lot of 👀

lewis6991 reviewed (approved):
Can we get an approval from another maintainer? @gpanders @mfussenegger @justinmk @MariaSolOs

MariaSolOs commented on runtime/doc/lsp.txt:
```diff
@@ -700,10 +701,30 @@ Lua module: vim.lsp                                                 *lsp-core*
+
+                         It's possibile to specify marker's priority. The
+                         default behaviour performs a directory-up first
+                         search, but it can be configured to look for multple
```
```suggestion
                         search, but it can be configured to look for multiple
```

MariaSolOs commented on runtime/doc/news.txt:
```diff
@@ -70,7 +70,24 @@ HIGHLIGHTS
 LSP
 
-• todo
+• `root_markers` in |vim.lsp.Config| can now specify markers priority.
+  Defaults to directory-up first search, which is more inline with
+  `nvim-lspconfig` but changes the default behaviour, which looked for every
+  marker before going direcotory-up.
```
```suggestion
  marker before going directory-up.
```

gpanders replied:
It's not strictly speaking true that each marker can "specify its priority", they are simply ordered by priority. Also there's no reason to mention nvim-lspconfig here.

```suggestion
• `root_markers` in |vim.lsp.Config| can now be ordered by priority.
```

gpanders commented on runtime/doc/news.txt:
```diff
@@ -70,7 +70,24 @@ HIGHLIGHTS
+  Defaults to directory-up first search, which is more inline with
+  `nvim-lspconfig` but changes the default behaviour, which looked for every
+  marker before going direcotory-up.
+  For example: >lua
+    root_markers = { 'stylua.toml', '.git' }
+<
+  Used to look for `stylua.toml` and `.git` in the current working directory
+  before going directory-up to repeat the same search.
+  Now this syntax looks for `stylua.toml` in every directory from the current
+  working directory going up, then it does the same for `.git`.
+
+  If you prefer the old behaviour, you need to specify that both markers have
+  the same priority, like this: >lua
+    root_markers = { { 'stylua.toml', '.git'} }
+<
+  For additional information (including mixed priorities), see `root_markers`
+  in |vim.lsp.Config|.
```
Too wordy for news. Direct users to the reference docs instead.

```suggestion
```

gpanders commented on runtime/lua/vim/lsp.lua:
```diff
@@ -283,7 +283,29 @@ end
 --- rootUri, and rootPath on initialization. Unused if `root_dir` is provided.
---- @field root_markers? string[]
+---
+--- It's possibile to specify marker's priority. The default behaviour performs a
+--- directory-up first search, but it can be configured to look for multple markers in each
+--- directory before going up.
```
We aren't specifying a marker's priority, we're allowing different markers to be ordered by priority. Maybe this is splitting hairs a bit, but when I read "specify a marker's priority" I'm expecting to see some kind of map where each marker can be assigned an arbitrary number, which isn't what this API provides.

```suggestion
--- Each entry in this list is a set of one or more markers. For each set, Nvim
--- will search upwards for each marker contained in the set. If a marker is
--- found, the directory which contains that marker is used as the root
--- directory. If no markers from the set are found, the process is repeated
--- with the next set in the list.
```

gpanders commented on runtime/lua/vim/lsp.lua:
````diff
@@ -283,7 +283,29 @@ end
+--- directory-up first search, but it can be configured to look for multple markers in each
+--- directory before going up.
+---
+--- Example (default, directory-up first):
+---
+--- ```lua
+---   root_markers = { 'stylua.toml', '.git' }
+--- ```
+---
+--- Looks for `stylua.toml` in every directory going up, if it isn't found it does
+--- the same for `.git`.
+---
+--- Example (mixed priorities):
+---
+--- ```lua
+---   root_markers = { { 'stylua.toml', '.luarc.json' }, '.git' }
+--- ```
+---
+--- Looks for both `.stylua.toml` and `.luarc.json` (in this order) in every
+--- directory going up, if none of them is found it does the same for `.git`.
+--- @field root_markers? (string|string[])[]
````
```suggestion
--- Example:
---
--- ```lua
---   root_markers = { 'stylua.toml', '.git' }
--- ```
---
--- Find the first parent directory containing the file `stylua.toml`. If not
--- found, find the first parent directory containing the file or directory
--- `.git`.
---
--- Example:
---
--- ```lua
---   root_markers = { { 'stylua.toml', '.luarc.json' }, '.git' }
--- ```
---
--- Find the first parent directory containing EITHER `stylua.toml` or
--- `.luarc.json`. If not found, find the first parent directory containing the
--- file or directory `.git`.
---
--- @field root_markers? (string|string[])[]
```

justinmk replied:
also helps to put the description in one place, not both above and below the example.

one way to do that is to put the explanation as a comment in the codeblock:

    Example:
    
    ```lua
      -- Find the first ...
      root_markers = ...
    ```

it is disorienting when examples are discussed both above and below each example. important details can be missed by the reader.

gpanders reviewed (commented):
Concept sounds good to me. I have some feedback on the documentation.

justinmk commented on runtime/lua/vim/lsp.lua:
```diff
@@ -651,7 +673,18 @@ function lsp.start(config, opts)
     config = vim.deepcopy(config)
-    config.root_dir = vim.fs.root(bufnr, opts._root_markers)
+
+    if type(opts._root_markers) == 'table' then
```
Is this condition ever false? I thought `opts._root_markers` is also a table when it's passed below (line 686). Did you mean this instead:
```suggestion
    if type(opts._root_markers[1]) == 'table' then
```

lewis6991 replied:
Good catch. I don't think this condition is necessary at all.

TheRealLorenz replied:
If that's the case I would add a `validate('root_markers', opts._root_markers, 'table')` just before.

TheRealLorenz replied:
I pushed a new approach.

lewis6991 replied:
`root_markers` is not a function, doesn't matter if it works.

justinmk commented on runtime/lua/vim/lsp.lua:
```diff
@@ -651,7 +673,18 @@ function lsp.start(config, opts)
-    config.root_dir = vim.fs.root(bufnr, opts._root_markers)
+
+    if type(opts._root_markers) == 'table' then
+      for _, marker in ipairs(opts._root_markers) do
+        local root = vim.fs.root(bufnr, marker)
```
If my above comment is correct, then this should be plural?
```suggestion
      for _, markers in ipairs(opts._root_markers) do
        local root = vim.fs.root(bufnr, markers)
```

lewis6991 replied:
Note `marker` is of type `string|string[]` here, so both are appropriate.

justinmk commented on runtime/lua/vim/lsp.lua:
```diff
@@ -651,7 +673,18 @@ function lsp.start(config, opts)
+          break
+        end
+      end
+    else
+      config.root_dir = vim.fs.root(bufnr, opts._root_markers)
```
could eliminate this `else` by changing the above to:

```lua
for _, marker in ipairs(type(opts._root_markers[1] == 'string' and { opts._root_markers } or opts._root_markers) do
  ...
```

lewis6991 replied:
I don't think we need the `else` at all.

justinmk commented on runtime/doc/lsp.txt:
```diff
@@ -700,10 +701,30 @@ Lua module: vim.lsp                                                 *lsp-core*
+                         if it isn't found it does the same for `.git`.
+
+                         Example (mixed priorities): >lua
+                               root_markers = { { 'stylua.toml', '.luarc.json' }, '.git' }
```
should we require the "nested" case to have all of its items as lists? else it complicates the logic.
```suggestion
                               root_markers = { { 'stylua.toml', '.luarc.json' }, { '.git' } }
```

lewis6991 replied:
Afaik it does not complicate the logic at all. We simply iterate root_markers and pass each elem (`string` or `string[]`) directly to `vim.fs.root`.

TheRealLorenz replied:
This is actually something that @lewis6991 suggested. https://github.com/neovim/neovim/pull/33485#discussion_r2055671928

justinmk replied:
Yeah true, since the `else` was [eliminated](https://github.com/neovim/neovim/pull/33485#discussion_r2056954917).

justinmk reviewed (approved):
Looks like a good approach. Needs a test.

TheRealLorenz commented:
Is it fine to add more fix up commits or should I rebase everything?

MariaSolOs commented:
> Is it fine to add more fix up commits or should I rebase everything?

You can just rebase to squash the fixups into the first 3 commits. See https://github.com/neovim/neovim/blob/master/CONTRIBUTING.md#commit-messages

justinmk commented on runtime/doc/lsp.txt:
```diff
@@ -41,6 +41,7 @@ Follow these steps to get LSP features:
         -- current buffer that contains either a ".luarc.json" or a
         -- ".luarc.jsonc" file. Files that share a root directory will reuse
         -- the connection to the same LSP server.
+        -- It also supports priorities see |vim.lsp.Config|
         root_markers = { '.luarc.json', '.luarc.jsonc' },
```
```suggestion
        -- Nested lists indicate equal priority at each "level".
        root_markers = { { '.luarc.json', '.luarc.jsonc' }, '.git' },
```

justinmk commented on runtime/doc/lsp.txt:
```diff
@@ -700,10 +701,39 @@ Lua module: vim.lsp                                                 *lsp-core*
+                         workspaceFolders, rootUri, and rootPath on
+                         initialization. Unused if `root_dir` is provided.
+
+                         It's possibile to specify marker's priority. The
```
This is misleading, because priority/precedence is always in effect. What we actually want to say is that you can opt-in to "equal precedence".
```suggestion
                         The list order decides the priority. To indicate "equal priority", specify names in a nested list.
```

TheRealLorenz replied:
My bad, I re-generated the docs without updating the lua file.

justinmk commented:
Will you be able to add a test? You can [run a single test](https://github.com/neovim/neovim/blob/master/test/README.md#filter-by-file) like this:

    TEST_FILE=test/functional/example_spec.lua make functionaltest

TheRealLorenz commented:
> Will you be able to add a test? You can [run a single test](https://github.com/neovim/neovim/blob/master/test/README.md#filter-by-file) like this:
> 
> ```
> TEST_FILE=test/functional/example_spec.lua make functionaltest
> ```

Wouldn't testing be easier if #33578 was tackled first?

justinmk commented:
> Wouldn't testing be easier if #33578 was tackled first?

Why is it blocked on that? There are existing `root_markers` tests here: https://github.com/neovim/neovim/blob/923efaea28ff019d2e864e15f3f00f6d6962dc94/test/functional/plugin/lsp_spec.lua#L6293

    TEST_FILE=test/functional/plugin/lsp_spec.lua TEST_FILTER=config TEST_TAG= make functionaltest

TheRealLorenz commented:
> > Wouldn't testing be easier if #33578 was tackled first?
> 
> Why is it blocked on that? There are existing `root_markers` tests here:
> 
> https://github.com/neovim/neovim/blob/923efaea28ff019d2e864e15f3f00f6d6962dc94/test/functional/plugin/lsp_spec.lua#L6293
> 
> ```
> TEST_FILE=test/functional/plugin/lsp_spec.lua TEST_FILTER=config TEST_TAG= make functionaltest
> ```

I'm not able to stop the client which connects to `_G.create_server()`, should I add any type of handler? I've already written the test, I just need to restart the client between test cases.

TheRealLorenz commented:
This is what I came up, but as mentioned the client is always there even after `stop`, so the root_dir is not updated.

```lua
      ---@param root_markers (string|string[])[]
      ---@param expected_root_dir string?
      local function markers_resolve_to(root_markers, expected_root_dir)
        exec_lua(function()
          vim.lsp.config['foo'] = {}
          vim.lsp.config('foo', {
            cmd = _G._custom_server.cmd,
            filetypes = { 'foo' },
            root_markers = root_markers,
          })
          vim.lsp.enable('foo')
          vim.cmd.edit(target)
          vim.bo.filetype = 'foo'
        end)
        retry(nil, 1000, function()
          eq(
            expected_root_dir,
            exec_lua(function()
              return vim.lsp.get_clients()[1].root_dir
            end)
          )
        end)
        exec_lua(function()
          vim.lsp.get_clients()[1]:stop(true)
          vim.lsp.get_clients()[1]:stop(true)
          vim.uv.sleep(1000)
          assert(#vim.lsp.get_clients() == 0) -- Added for debugging FAILS
        end)
      end

```

Where
```lua
_G._custom_server = _G._create_server()
```

justinmk commented:
> the client is always there even after `stop`

maybe `reuse_client=function() return false end` could help.

but maybe the server needs to be stopped.

TheRealLorenz committed:
test(lsp): add testing for root_markers

TheRealLorenz commented:
> > the client is always there even after `stop`
> 
> 
> 
> maybe `reuse_client=function() return false end` could help.
> 
> 
> 
> but maybe the server needs to be stopped.

I used `reuse_client = function() return false end` and by fetching the last client it works. I also rebased with current master.

justinmk commented:
> the client is always there even after `stop`

maybe related:

- https://github.com/neovim/neovim/issues/33616

justinmk reviewed (approved):
Nicely done! The test seems fast too. I will tweak the docs a bit.

I don't see any breaking changes, so I removed "!" from the commit subject.

justinmk commented:
> `root_markers` ... actually also supports `string` and `function(name: string, path: string)`, but they're not documented

Are those actually needed? Isn't it redundant with `root_dir` ?

TheRealLorenz commented:
> > `root_markers` ... actually also supports `string` and `function(name: string, path: string)`, but they're not documented
> 
> 
> 
> Are those actually needed? Isn't it redundant with `root_dir` ?

Yes,they mostly are. It was relevant because it made searching for patterns a little bit less cumbersome, but it requires some modification in core.
I also put a `vim.validate` which requires it to be a `table` only.

It was supposed to be a breaking change because it changes the default behaviour, the same table passed as root markers before was searched pattern-first and now directory-up first.

lithammer commented:
Would it make sense expose this as a (new) function (e.g. `vim.fs.root_markers`) or add an option to `vim.fs.root()`? Or just plain make a breaking change since `vim.fs.root()` is more or less made for this exact use-case if I understand correctly?

Currently it's a bit awkward since you have to go from this:

```lua
root_markers = { 'prio1.json', 'prio2.json', 'prio3.json' },
```

To this:

```lua
root_dir = function(bufnr, on_dir)
  on_dir(
    vim.fs.root(bufnr, 'prio1.json') or
    vim.fs.root(bufnr, 'prio2.json') or
    vim.fs.root(bufnr, 'prio3.json')
  )
end,
```

I expect some configs are incorrect now because they've gone from [`lspconfig.util.root_pattern()`](https://github.com/neovim/nvim-lspconfig/blob/61e5109c8cf24807e4ae29813a3a82b31821dd45/lua/lspconfig/util.lua#L30-L48) to `root_dir` and `vim.fs.root()`. Basically they gone from caring about the order of the markers to not caring.

One such example is `gopls` which now has this which is incorrect since `go.work` should take priority over `go.mod` (which it does in the legacy config).

```lua
vim.fs.root(fname, { 'go.work', 'go.mod', '.git' })
```

https://github.com/neovim/nvim-lspconfig/blob/61e5109c8cf24807e4ae29813a3a82b31821dd45/lsp/gopls.lua#L18

TheRealLorenz commented:
> Currently it's a bit awkward since you have to go from this:
> 
> ```lua
> root_markers = { 'prio1.json', 'prio2.json', 'prio3.json' },
> ```
> 
> To this:
> 
> ```lua
> root_dir = function(bufnr, on_dir)
>   on_dir(
>     vim.fs.root(bufnr, 'prio1.json') or
>     vim.fs.root(bufnr, 'prio2.json') or
>     vim.fs.root(bufnr, 'prio3.json')
>   )
> end,
> ```

You don't, the first snippet works as intended in neovim 0.11.1+. The purpose of the PR was to align `root_markers` with `util.root_pattern()` to correctly handle priorities. It also provides a way to have "same priority" markers but it's opt-in.

> One such example is `gopls` which now has this which is incorrect since `go.work` should take priority over `go.mod` (which it does in the legacy config).
> 
> ```lua
> vim.fs.root(fname, { 'go.work', 'go.mod', '.git' })
> ```
> 
> https://github.com/neovim/nvim-lspconfig/blob/61e5109c8cf24807e4ae29813a3a82b31821dd45/lsp/gopls.lua#L18
> 

While there are cases where custom logic (like `root_dir`) might require a more verbose approach, these are relatively rare, which doesn't justify a change to the core behaviour of `vim.fs.root()`.

lithammer commented:
> You don't, the first snippet works as intended in neovim 0.11.2+.

My example was perhaps bad, but I wanted to keep it short for brevity. I meant that you would go to something like this:

```diff
 root_dir = function(bufnr, on_dir)
   on_dir(
+    some_custom_logic() or
     vim.fs.root(bufnr, 'prio1.json') or
     vim.fs.root(bufnr, 'prio2.json') or
     vim.fs.root(bufnr, 'prio3.json')
   )
 end,
```

> The purpose of the PR was to align `root_markers` with `util.root_pattern()` to correctly handle priorities. It also provides a way to have "same priority" markers but it's opt-in.

Right. And I'm saying `vim.fs.root()` should get the same treatment. Or at least a sibling function.

The expectation from me at least is that `root_markers = {"foo", "bar", "baz"}` is just a nicer way of writing `vim.fs.root(bufnr, {"foo", "bar", "baz"})`. What other purpose does it have?

> While there are cases where custom logic (like `root_dir`) might require a more verbose approach, these are relatively rare, which doesn't justify a change to the core behaviour of `vim.fs.root()`.

I'm not convinced. What if the core behaviour of `vim.fs.root()` is incorrect/unwanted as evident by this very PR (and `util.root_pattern()`)?

justinmk commented:
> I'm saying `vim.fs.root()` should get the same treatment

Agreed. Tracked in https://github.com/neovim/neovim/issues/34258 , pr very welcome
